runtime/src/iree/modules/vmvx/module.c

```diff
@@ -61,21 +61,431 @@ iree_vmvx_module_free_state(void* self, iree_vm_module_state_t* module_state) {
+    }
+  }
+
+  IREE_TRACE_ZONE_END(z0);
+  return iree_ok_status();
+}
+
+IREE_VM_ABI_EXPORT(iree_vmvx_matmul_f32f32f32, iree_vmvx_module_state_t,
+                   rIIrIIrIIIIIffi, v) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  MAP_BUFFER_2D_RO(lhs, float,
+                   /*buffer_ref=*/
+                   args->matmul_f32.lhs_ref,
+                   /*offset=*/
+                   args->matmul_f32.lhs_offset,
+                   /*stride0=*/
+                   args->matmul_f32.lhs_row_stride,
+                   /*stride1=*/
+                   1,
+                   /*size0=*/
+                   args->matmul_f32.m,
+                   /*size1=*/
+                   args->matmul_f32.k);
+  MAP_BUFFER_2D_RO(rhs, float,
+                   /*buffer_ref=*/
+                   args->matmul_f32.rhs_ref,
+                   /*offset=*/
+                   args->matmul_f32.rhs_offset,
+                   /*stride0=*/
+                   args->matmul_f32.rhs_row_stride,
+                   /*stride1=*/
+                   1,
+                   /*size0=*/
+                   args->matmul_f32.k,
+                   /*size1=*/
+                   args->matmul_f32.n);
+  MAP_BUFFER_2D_RW(out, float,
+                   /*buffer_ref=*/
+                   args->matmul_f32.out_ref,
+                   /*offset=*/
+                   args->matmul_f32.out_offset,
+                   /*stride0=*/
+                   args->matmul_f32.out_row_stride,
+                   /*stride1=*/
+                   1,
+                   /*size0=*/
+                   args->matmul_f32.m,
+                   /*size1=*/
+                   args->matmul_f32.n);
+
+  iree_host_size_t M = (iree_host_size_t)args->matmul_f32.m;
+  iree_host_size_t N = (iree_host_size_t)args->matmul_f32.n;
+  iree_host_size_t K = (iree_host_size_t)args->matmul_f32.k;
+
+  // TODO: Define
+  // flags more
+  // robustly.
+  if (args->matmul_f32.flags == 0) {
+    // Row major.
+    for (iree_device_size_t i = 0; i < M; ++i) {
+      for (iree_device_size_t k = 0; k < K; ++k) {
+        float apart = args->matmul_f32.alpha * lhs[i * lhs_stride0 + k];
+        for (iree_device_size_t j = 0; j < N; ++j) {
+          out[i * out_stride0 + j] +=
+              args->matmul_f32.beta * apart * rhs[k * rhs_stride0 + j];
+        }
+      }
+    }
+  } else {
+    IREE_TRACE_ZONE_END(z0);
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "unsupported matmul flags: %x",
+                            (unsigned)args->matmul_f32.flags);
+  }
+
+  IREE_TRACE_ZONE_END(z0);
   return iree_ok_status();
 }
 
 //===----------------------------------------------------------------------===//
-// VM module interface implementation
+// VM module
+// interface
+// implementation
 //===----------------------------------------------------------------------===//
 
-// NOTE: this must match the ordering of the iree_vmvx_module_exports_ table.
+// NOTE: this must
```

Comment:
looks like some auto formatting thing?

Reply:
Yeah. clang-formatted with a missing backslash in a macro... and it squished the remainder of the file into the last ten columns. Then I didn't notice it had mangled everything.